### PR TITLE
Feature/abd storytime

### LINKED ISSUE #57: Should all contributors be able to see the original prompt?

### PR DESCRIPTION
**Pull request by Abdeboskey**
#### What's this PR do?
* Adds the ability to post and patch stories
* Refactored a bunch o things on the way to hook these components and methods up
#### Where should the reviewer start?
* Start at ApiHelper.js, and then peek at StoryEditFooter.js
#### How should this be manually tested?
* Pull this branch, and attempt to contribute to a story or start a new story. 
#### Any background context you want to provide?
* When doing this, we were having real tricky issues passing methods through the redirect in StorySetupView and WelcomeView, so we found a work around by passing them directly to props for the StoryEdit path in App.js
#### What are the relevant tickets?
NA Free Willy
#### Screenshots (if appropriate)
free willy
#### Questions/Next steps:
* Create option to select a new prompt if you don't like a suggestion
* Should all contributors be able to see the original prompt?
* Fix bookshelf hovers
* Username should not be case sensitive for email or username on the backend
* Proptypes !!!!
* create error message for incorrect password on login
* Timer should begin when user begins typing in story input

### CODE REVIEW

**Review by nickhartdev — APPROVED** [review]
Appreciate y'all's work on this! I'm glad we learned our lesson about passing props through a redirect - it seemed convenient at first but with the changes y'all put in now things seem much cleaner.

**Comment by Abdeboskey** [comment]
@nickhartdev fo sho. I believe it's still ok to pass data through a redirect, but passing methods that use `.this` seemed to be where it got mad.

File: src/ApiHelper/ApiHelper.js [hunk 1]
```diff
@@ -9,14 +9,27 @@ class ApiHelper {
...
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify(loginInfo),
+      })
+    } catch (error) {
+      return error
+    }
+  }
```
Comment (carlymclift) [inline src/ApiHelper/ApiHelper.js:22]: Thanks for adding this catch error!

File: src/components/Login/Login.js [hunk 2]
```diff
@@ -30,28 +34,33 @@ class Login extends Component {
...
-
-     console.log(this.state)
+    if (redirect) {
+      return <Redirect to='/'/>;
+    }
 
     return (
       <div className="Login-container">
-        <form className="Login-form">
+        <form className="Login-form" onSubmit={this.verifyLogin}>
```
Comment (carlymclift) [inline src/components/Login/Login.js:63]: I forgot about onSubmit! thank you!

File: src/components/StoryEditFooter/StoryEditFooter.js [hunk 3]
```diff
@@ -1,26 +1,58 @@
...
+      contributions: textInputs.story,
+      contributors: author.id
+    }
+    if (!story.id) {
+      body.title = textInputs.title
+      body.prompt = story.prompt.id
+    }
+    if (isFinished) body.is_complete = true
+    return body
+  }
```
Comment (carlymclift) [inline src/components/StoryEditFooter/StoryEditFooter.js:20]: BodyBuilder!!!

**Review by carlymclift — APPROVED** [review]
Thanks for all these changes1

**Comment by GreysonElkins** [comment]
I'd like to add that passing methods which `setState` through a redirect will change the scope of `this` to affect `History.location`s state instead of the original component. Something to watch out for!

File: src/components/App/App.js [hunk 4]
```diff
@@ -92,6 +91,13 @@ class App extends Component {
     return this.state.stories.filter(story => story.is_complete)
   }
 
+  addStory = (story) => {
+    const newStories = [...this.state.stories, story]
```
Comment (GreysonElkins) [inline src/components/App/App.js:95]: I love this concat method so much. Where's it been all my life?

File: src/components/App/App.js [hunk 5]
```diff
@@ -113,8 +119,11 @@ class App extends Component {
           render={() => {
             return (
               <WelcomePageView
+                author={this.state.currentUser}
                 stories={this.incompleteStories()}
                 authorUpdater={this.updateContributorData}
+                addStory={this.addStory}
+                updateStoryData={this.updateStoryData}
```
Comment (GreysonElkins) [inline src/components/App/App.js:126]: These two lines can go now, yeah?

Comment (GreysonElkins) [reply src/components/App/App.js:126]: And 122?

File: src/components/App/App.js [hunk 6]
```diff
@@ -123,13 +132,23 @@ class App extends Component {
           exact
           path="/story-setup"
           render={() => {
-            return <StorySetupView prompts={this.state.prompts} />;
+            return (
+              <StorySetupView
+                prompts={this.state.prompts}
+                author={this.state.currentUser}
+                addStory={this.addStory}
+                updateStoryData={this.updateStoryData}
```
Comment (GreysonElkins) [inline src/components/App/App.js:140]: Last three lines necessary?

File: src/components/App/App.js [hunk 7]
```diff
@@ -123,13 +132,23 @@ class App extends Component {
...
+            );
           }}
         />
         <Route
           exact
           path="/story-edit"
           render={(props) => {
+            props.updateStoryData = this.updateStoryData
+            props.addStory = this.addStory
+            props.author = this.state.currentUser
```
Comment (GreysonElkins) [inline src/components/App/App.js:151]: _props_ the **_object_** y'all. So cool

File: src/components/Bookshelf/Bookshelf.scss [hunk 8]
```diff
@@ -1,22 +1,18 @@
 .bookContainer {
   display: flex;
-  justify-content: center;
+  justify-content: flex-start;
```
Comment (GreysonElkins) [inline src/components/Bookshelf/Bookshelf.scss:3]: I'm still not sold on the left-to-right thing.
Can we have this change based on some mixin, or parameter? If the total width of the books exceeds the width of the container or something?

File: src/components/Bookshelf/Bookshelf.scss [hunk 9]
```diff
@@ -1,22 +1,18 @@
...
   background-size: 2em;
   background-repeat: no-repeat;
-  color: white;
-  display: flex; 
-  flex-direction: column;
   height: 10em;
   padding: .01em;
-  text-decoration: none;
-  width: 2em;
+  min-width: 2em;
```
Comment (GreysonElkins) [inline src/components/Bookshelf/Bookshelf.scss:15]: Thanks for pushing through to this!

File: src/components/StoryEditFooter/StoryEditFooter.js (hunk 3 above)
Comment (GreysonElkins) [reply src/components/StoryEditFooter/StoryEditFooter.js:20]: Lol I love the body builder; I def think we're gonna get marked for naming conventions tho 😂

File: src/components/StoryEditFooter/StoryEditFooter.js [hunk 10]
```diff
@@ -1,26 +1,58 @@
...
       <TimerDisplay disableStoryInput={disableStoryInput} />
       <Link to="/">
-        <button type="button" id="post-button">
-          Pass it on
+        <button 
+          type="button" 
+          id="post-button" 
+          onClick={postStory}
+        >
+          {!story.id ? "Continue Story" : "Start Story"}
```
Comment (GreysonElkins) [inline src/components/StoryEditFooter/StoryEditFooter.js]: These should get flipped

Comment (GreysonElkins) [reply src/components/StoryEditFooter/StoryEditFooter.js]: I may pop in right here and change that.

**Review by GreysonElkins — APPROVED** [review]
Great work bud!

**Pull request merged**